### ISSUE TITLE
thread channel becomes null when view_channel is restricted in parent text channel

### ISSUE THREAD

**Issue by pkdev08** [discordjs/discord.js#7883]
### Which package is this bug report for?

discord.js

### Issue description

Issue: Thread channel becomes null when "View_Channel" permissions are removed from parent text channel if client is in thread.

Steps to reproduce:
1) Create a thread under a text channel
2) Have the bot client join this thread
3) Restrict "View_Channel" permissions in the parent text channel
4) The thread channel becomes null, although the client is still in there.
5) If you try to use an interaction or such, it will error as you can't check client permissions on a null channel

Expected Behavior:
The thread channel should remain in the cache, so you can check permissions such as thread.sendable and thread.viewable

<img width="162" alt="Screen Shot 2022-05-04 at 6 59 35 PM" src="https://user-images.githubusercontent.com/54294685/166838808-f1a78939-1ffa-4079-a834-6371b2098226.png">

### Code sample

```typescript
N/A
```

### Package version

discord.js@13.6.0

### Node.js version

v16.15.0

### Operating system

Debian

### Priority this issue should have

High (immediate attention needed)

### Which partials do you have configured?

User, Channel, GuildMember, Message, Reaction

### Which gateway intents are you subscribing to?

Guilds, GuildMembers, GuildInvites, GuildVoiceStates, GuildMessages, GuildMessageReactions, DirectMessages

**Comment by pkdev08** [comment]
Once you toggle View Channel off, you might have to reboot the bot, and then the channel won't be cached. Any interaction that requires you check the client permissions, would fail. or any events sending messages to these channels. Make sure the bot does not have Admin when re-producing this issue.

**Comment by Jiralite** [comment]
Why is this a discord.js issue?

**Comment by pkdev08** [comment]
.. Because the channel becomes null?

**Comment by pkdev08** [comment]
It seems as the thread channel is not cached if the client doesn't have permissions in it.

**Comment by Jiralite** [comment]
> .. Because the channel becomes null?

Is this not expected behaviour? You cannot resolve the channel, thus, it is `null`?

**Comment by pkdev08** [comment]
Either why the the interactions are still registered in the channel without client having access, and therefore you can't run any permission checks.

**Comment by pkdev08** [comment]
Why is a normal text channel still cached even if the client does not have access to view it?

**Comment by ckohen** [comment]
TLDR: behaving as (better than?) expected.

<details><summary>More detailed explanation on the whys and hows></summary>
That is a design decision from Discord (hopefully changing for channels too at some point in the future). When threads were created, there was explicit intent given to ensure that threads a client is not part of are entirely restricted from access. You can read more about that [here](https://discord.com/developers/docs/topics/threads#additional-context-on-the-the-threadlistsync-and-threadcreate-dispatches).

Ever notice how mentioning a channel you can't see in a guild will still show the channel name, but a thread you can't view shows as deleted channel? That's exactly what's happening with your bot here. Your bot is still a *member* of the thread, but it is no longer *in* the thread, that is what the crossed out member means.

Your *application* can still receive interactions from that channel because applications don't abide by any role permissions, only utilizing command permissions.

Now, the question of why I said "better than" in the TLDR. What's interesting here is that the thread is actually being removed from cache at all. This *is* the intended, correct, behavior, but there is nothing currently written in discord.js to remove threads you no longer have access to from cache, aside from restarting (brings to light a memory leak, thank you for that). That is because Discord doesn't actually send an event specifically letting us know that your bot no longer has access, we only get a relevant permissions update via some other event. Handling every edge case of that is damn well near impossible, so we don't.

So now you may ask why its better that it be removed from cache rather than remaining present. That's quite simple. Since Discord no longer sends events regarding those threads, your cache quickly falls out of date. You may happen to luck out with permission checks since they rely on the parent channel, which will still receive updates. However, its better to just treat it like a thread you don't have access to, because yes, you can most certainly receive interactions from threads that you've never gotten a single piece of information about. If you can't handle a thread being null after previously seeing it, how do you handle threads you've never seen?
</details>

**Comment by pkdev08** [comment]
Hmm interesting. Thanks for the explanation. I guess I need to look into this for a fix, I've never really had this issue arise in production, although I stumbled across it during some development tests. I guess there is ThreadChannel.viewable, but not sure what other possible way there would be to check. Some of my interactions use other permissions such as MANAGE_MESSAGES, etc, etc, for some of them it shouldn't be an issue, but one's where I need to check specific perms, this will throw errors.

**Comment by pkdev08** [comment]
So really the issue at hand is if the client can't view a channel, since there's no channel in the cache, you can't even run ThreadChannel.viewable or anything, how else could you check that the client can't view the channel?

**Comment by ckohen** [comment]
If the channel for an interaction is null it means one of two things. 

1. The interaction is from a thread that the bot is not able to view OR
2. The interactions is from a guild only authorized with `applications.commands`

In both cases, null is an accurate representation of the channel, and you can assume 0 permissions, because that is what you have. The API will not let you do *anything* in those channels outside of the normal interaction flow.

**Comment by almostSouji** [comment]
This is not within our power to change. Especially if you need to reboot the bot to reproduce the issue, the bot will never receive the thread channel and can thus not cache it, thus we cannot retrieve it from the cache.

**Comment by pkdev08** [comment]
Thank you all for the help. I will just assume a null channel implies 0 permissions. Thanks
